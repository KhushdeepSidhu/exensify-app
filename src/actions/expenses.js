import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = ( expense ) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = ( expenseData = {} ) => {
    return ( dispatch, getState ) => {
        const uid = getState().auth.uid
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        return database.ref( `users/${uid}/expenses` ).push ( expense ).then ( ( ref ) => {
            dispatch ( addExpense ( {
                id: ref.key,
                ...expense
            }) )
        } )

    }
}

// EDIT_EXPENSE
export const editExpense = ( { description, note, amount, createdAt }, id ) => {
    return ( {
        type: 'EDIT_EXPENSE',
        expense: {
            id,
            description,
            note,
            amount,
            createdAt
        }
    
} ) }

export const startEditExpense = ( { description, note, amount, createdAt }, id ) => {
    return ( dispatch, getState ) => {
        const uid = getState ().auth.uid
        return database.ref ( `users/${uid}/expenses/${id}` ).update ( { description, note, amount, createdAt } )
        .then ( () => {
            dispatch ( editExpense ( { description, note, amount, createdAt }, id ) )
        } )
    } 
}


// REMOVE_EXPENSE
export const removeExpense = ( id ) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
})

// startRemoveExpense
export const startRemoveExpense = ( id ) => {
    return ( dispatch, getState ) => {
        const uid = getState().auth.uid
        return database.ref ( `users/${uid}/expenses/${id}` ).remove ().then ( () => {
            dispatch ( removeExpense ( id ) )
        } )
    }
}

// SET_EXPENSE
export const setExpenses = ( expenses ) => ( {
    type: 'SET_EXPENSES',
    expenses
} )

//export const startSetExpense
export const startSetExpenses = () => {
    return ( dispatch, getState ) => {
        const uid = getState().auth.uid
        return database.ref( `users/${uid}/expenses` ).once ( 'value' ).then ( ( snapshot ) => {
            
            const expenses = []

            snapshot.forEach ( ( childSnapshot ) => {
                expenses.push ( {
                id: childSnapshot.key,
                ...childSnapshot.val()
            } )
        })
        dispatch ( setExpenses ( expenses ) )
        } )
    }
}