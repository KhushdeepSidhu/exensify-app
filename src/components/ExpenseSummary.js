import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectedExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'

export const ExpenseSummary = ( props ) => {
    const expenseCount = props.expenses.length
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const formattedExpensesTotal = numeral ( getExpensesTotal ( props.expenses ) / 100 ).format ('$0,0.00')
    return (
        <div>
            <p>
                { `Viewing ${ expenseCount } ${ expenseWord } totalling ${ formattedExpensesTotal }` }
            </p>
        </div>
    )
}


const mapStatetoProps = ( state ) => {
    return {
        expenses: selectedExpenses ( state.expenses, state.filters )
    }
}

export default connect ( mapStatetoProps ) ( ExpenseSummary )
