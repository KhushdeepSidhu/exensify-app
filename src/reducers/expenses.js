// Expenses Reducer
const expenseReducerDefaultState = []

export default ( state = expenseReducerDefaultState, action ) => {
    switch ( action.type ) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ]

        case 'REMOVE_EXPENSE':
            return state.filter( ( expense ) => expense.id !== action.expense.id )

        case 'EDIT_EXPENSE':
            return state.map ( ( expense ) => {
                if ( expense.id === action.expense.id ) {
                    return {
                        ...expense,
                        ...action.expense
                    }
                } else {
                    return expense
                }
            } )

        case 'SET_EXPENSES':
            return action.expenses
            
        default: 
            return state

    }
}