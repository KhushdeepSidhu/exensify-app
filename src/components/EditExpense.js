import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpense extends React.Component {

    onSubmit = ( expense ) => {
        this.props.startEditExpense ( expense, this.props.expense.id ) 
        this.props.history.push ( '/' )
    }

    onRemove = () => {
        this.props.startRemoveExpense ( this.props.expense.id  )
        this.props.history.push ( '/' )
    }

    render () {
        return (
            <div>
                <ExpenseForm
                    expense={ this.props.expense } 
                    onSubmit = { this.onSubmit }
                /> 
                <button
                    onClick = { this.onRemove }>
                    Remove
                </button>
            </div>
        )
    }

}

const mapStateToProps = ( state, props ) => {
    return {
        expense: state.expenses.find ( ( expense ) => expense.id === props.match.params.id )
    }
}

const mapDispatchToProps = ( dispatch ) => ( {
    startEditExpense: ( expense, id ) => dispatch ( startEditExpense ( expense, id ) ),
    startRemoveExpense: ( id ) => dispatch ( startRemoveExpense ( id ) )
} )

export default connect ( mapStateToProps, mapDispatchToProps ) ( EditExpense )