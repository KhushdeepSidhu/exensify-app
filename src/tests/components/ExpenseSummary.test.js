import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses'

test ( 'Should render expense summary correctly with multiple expenses', () => {
    const wrapper = shallow ( <ExpenseSummary expenses = { expenses }/> )
    expect ( wrapper ).toMatchSnapshot()
} )

test ( 'Should render expense summary correctly with single expense', () => {
    const wrapper = shallow ( <ExpenseSummary expenses = { [ expenses[1] ] }/> )
    expect ( wrapper ).toMatchSnapshot()
} )