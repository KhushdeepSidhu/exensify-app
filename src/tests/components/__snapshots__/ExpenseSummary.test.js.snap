// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense summary correctly with multiple expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $1,143.45
  </p>
</div>
`;

exports[`Should render expense summary correctly with single expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1,095.00
  </p>
</div>
`;
